In [2]:
import pandas as pd

### 1. Read in the data 

In [3]:
df = pd.read_csv("NYC Open Data - Electric_Consumption_And_Cost__2010_-__June_2018_.csv")

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
df.shape

(313147, 27)

#### Check the number of empty values in each column

In [5]:
df.isnull().sum()

Development Name         146
Borough                  146
Account Name             146
Location                9041
Meter AMR                187
Meter Scope           296588
TDS #                   1717
EDP                      146
RC Code                  146
Funding Source           146
AMP #                   1657
Vendor Name              146
UMIS BILL ID             146
Revenue Month            146
Service Start Date       146
Service End Date         146
# days                   146
Meter Number             146
Estimated                146
Current Charges          146
Rate Class               146
Bill Analyzed            146
Consumption (KWH)        146
KWH Charges              146
Consumption (KW)         146
KW Charges               146
Other charges            146
dtype: int64

### 2. Remove empty rows

In [5]:
mask = (pd.isna(df['Account Name']) == True)
df.drop(mask[mask == True].index, axis = 0, inplace = True)

### 3. Remove rows where electricity charges were estimated

In [6]:
df.drop(df.loc[df['Estimated'] == 'Y         '].index, axis = 0, inplace = True)

#### Check data types of columns

In [7]:
df.dtypes

Development Name       object
Borough                object
Account Name           object
Location               object
Meter AMR              object
Meter Scope            object
TDS #                 float64
EDP                   float64
RC Code                object
Funding Source         object
AMP #                  object
Vendor Name            object
UMIS BILL ID          float64
Revenue Month          object
Service Start Date     object
Service End Date       object
# days                float64
Meter Number           object
Estimated              object
Current Charges        object
Rate Class             object
Bill Analyzed          object
Consumption (KWH)     float64
KWH Charges            object
Consumption (KW)       object
KW Charges             object
Other charges          object
dtype: object

### 4. Change the following fields from string to float

"Consumption (KW)", "Current Charges", "KWH Charges", "KW Charges", "Other charges"

In [8]:
df["Consumption (KW)"] = df["Consumption (KW)"].apply(lambda x: x.replace(",","") if type(x) == str else str(x))
df["Consumption (KW)"] = df["Consumption (KW)"].astype(float)

In [9]:
df["Current Charges"] = df["Current Charges"].apply(lambda x: x.replace("$","").replace(",","").replace("(","-").replace(")","") if type(x) == str else str(x))
df["Current Charges"] = df["Current Charges"].astype(float)

In [10]:
df["KWH Charges"] = df["KWH Charges"].apply(lambda x: x.replace("$","").replace(",","").replace("(","-").replace(")","") if type(x) == str else str(x))
df["KWH Charges"] = df["KWH Charges"].astype(float, inplace = True)

In [11]:
df["KW Charges"] = df["KW Charges"].apply(lambda x: x.replace("$","").replace(",","").replace("(","-").replace(")","") if type(x) == str else str(x))
df["KW Charges"] = df["KW Charges"].astype(float, inplace = True)

In [12]:
df["Other charges"] = df["Other charges"].apply(lambda x: x.replace("$","").replace(",","").replace("(","-").replace(")","") if type(x) == str else str(x))
df["Other charges"] = df["Other charges"].astype(float, inplace = True)

#### Summarize the numerical fields

More than 25% of the values for all except "Curent Charges" are 0

In [13]:
df[["Consumption (KWH)",  "Consumption (KW)", "Current Charges", "KWH Charges", "KW Charges", "Other charges"]].describe()

Consumption (KWH)  Consumption (KW)  Current Charges    KWH Charges  \
count       2.612520e+05     261252.000000    261252.000000  261252.000000   
mean        3.256572e+04         68.239881      4510.758176    1673.941416   
std         5.307171e+04        122.266314      6630.228947    2921.443192   
min         0.000000e+00          0.000000      -243.150000       0.000000   
25%         0.000000e+00          0.000000       384.342500       0.000000   
50%         1.136000e+04         31.510000      2536.945000     555.020000   
75%         4.816000e+04         98.500000      6052.357500    2359.892500   
max         1.779600e+06      16135.460000    329800.370000  195575.860000   

          KW Charges  Other charges  
count  261252.000000  261252.000000  
mean     1084.801607    1672.128898  
std      1807.808563    3626.935491  
min         0.000000  -59396.430000  
25%         0.000000       0.000000  
50%       448.370000     896.510000  
75%      1601.422500    2636.255000  
max     78782.960000  134224.510000

### 5. Convert Revenue Month and Two dates to datetime type

In [14]:
from datetime import datetime

In [15]:
df["Revenue Month"] = df["Revenue Month"].map(lambda x: datetime.strptime(x.split(" ")[0], '%m/%d/%Y'))
df['Service Start Date'] = df['Service Start Date'].map(lambda x: datetime.strptime(x, '%m/%d/%Y'))
df['Service End Date'] = df['Service End Date'].map(lambda x: datetime.strptime(x, '%m/%d/%Y'))

#### Check "Meter Scope": are the row with a range value represent a "Master Meter" (i.e. its value is the sum of other related rows)? - No

In [16]:
df['Meter Scope'].value_counts()

Community Center                                      3223
BLD 01                                                 621
BLD 02                                                 452
BLD 03                                                 433
BLD 04                                                 402
BLD 06                                                 274
BLD 07                                                 274
APT 01A - Management Office                            273
APT 01B - Management Office                            263
WEST BRIGHTON I ALL                                    200
WEST BRIGHTON II ALL                                   200
APT 02F - Management Office                            195
APT 02E - Management Office                            194
MORRISANIA ALL                                         194
APT 02H - Management Office                            194
APT 02G - Management Office                            193
BLD 05                                                 1

In [17]:
df[(df['TDS #'] == 118) & (df["Revenue Month"] == '2010-02-01')][["Location", "Meter Scope", "Revenue Month", "Current Charges"]]

Location   Meter Scope Revenue Month  Current Charges
1    BLD 05  BLD 01 to 07    2010-02-01         14556.34
13   BLD 06        BLD 06    2010-02-01         10948.54
25   BLD 07        BLD 07    2010-02-01         10403.82
37   BLD 01        BLD 01    2010-02-01          9422.06
49   BLD 02        BLD 02    2010-02-01          9710.06
61   BLD 03        BLD 03    2010-02-01          9286.81
73   BLD 04        BLD 04    2010-02-01         10133.33

In [18]:
df[(df['Meter Scope'] == 'Community Center')].groupby('Location').mean()

TDS #         EDP  UMIS BILL ID     # days  \
Location                                                                     
BLD 01 - Community Center  219.172222  448.761111  3.703092e+06  30.455556   
BLD 02 - Community Center  179.483471  435.892562  3.678972e+06  30.342975   
BLD 03 - Community Center  227.368078  345.739414  3.657077e+06  30.358306   
BLD 04 - Community Center  164.564103  458.384615  3.592468e+06  30.318681   
BLD 05 - Community Center  111.000000  434.000000  3.727165e+06  30.351064   
BLD 06 - Community Center  238.207447  457.611702  3.646114e+06  30.367021   
BLD 08                      75.000000  226.000000  3.734546e+06  30.474747   
BLD 09 - Community Center   28.000000  523.000000  3.892687e+06  30.421687   
BLD 10 - Community Center   60.928571  392.952381  3.626191e+06  30.425926   
BLD 12 - Community Center  126.000000  440.000000  4.053056e+06  30.463415   
BLD 13 - Community Center   88.000000  234.000000  2.897605e+06  30.285714   
BLD 15 - Community Center   44.000000  584.000000  3.702243e+06  30.416667   
BLD 16 - Community Center  243.000000  353.000000  2.926861e+06  30.321429   
BLD 21 - Community Center    2.000000  200.000000  2.865459e+06  30.561404   
BLD 23 - Community Center   89.000000  235.000000  3.640108e+06  30.382979   
BLD 25 - Community Center   92.000000  670.000000  3.630029e+06  30.438776   
BLD 31 - Community Center   67.000000  222.000000  3.858227e+06  30.377778   
BLD 39 - Community Center  514.000000  509.000000  2.890956e+06  30.250000   
BLD 48 - Community Center  377.000000  650.000000  3.706621e+06  30.376289   
BLD 51 - Community Center  303.000000  566.000000  3.944821e+06  30.233766   

                           Current Charges  Consumption (KWH)  KWH Charges  \
Location                                                                     
BLD 01 - Community Center       968.792556        5736.194444   313.216667   
BLD 02 - Community Center      1786.717975        9327.508264   509.086653   
BLD 03 - Community Center      3364.966645       22792.403909  1209.733257   
BLD 04 - Community Center      2934.512381       17003.956044   922.177802   
BLD 05 - Community Center      1359.712021        9234.893617   492.193723   
BLD 06 - Community Center      3094.852234       17753.829787   954.731543   
BLD 08                         2617.118889       12876.363636   641.169293   
BLD 09 - Community Center      4268.201084       23331.084337  1204.569036   
BLD 10 - Community Center      4550.163545       26973.185185  1412.216614   
BLD 12 - Community Center      4852.131341       22640.975610  1194.572927   
BLD 13 - Community Center      1350.259796        8073.469388   470.809592   
BLD 15 - Community Center      1580.647396        9593.750000   518.277917   
BLD 16 - Community Center      3858.404107       17451.428571   994.812500   
BLD 21 - Community Center      4539.154561       23416.140351  1344.025439   
BLD 23 - Community Center      2269.757234       14081.702128   752.744255   
BLD 25 - Community Center      4674.129286       30216.326531  1597.667857   
BLD 31 - Community Center      6516.316667       40022.222222  2111.066667   
BLD 39 - Community Center      1594.283594        5899.375000   343.410469   
BLD 48 - Community Center      1965.927887       10706.391753   567.904072   
BLD 51 - Community Center      2001.323896       12703.896104   649.832987   

                           Consumption (KW)   KW Charges  Other charges  
Location                                                                 
BLD 01 - Community Center         16.535889   216.624500     423.685889  
BLD 02 - Community Center         33.722045   400.787686     833.356219  
BLD 03 - Community Center         55.999088   675.012866    1395.811466  
BLD 04 - Community Center         50.013919   712.859634    1244.547033  
BLD 05 - Community Center         22.808511   278.162553     559.105426  
BLD 06 - Community Center         51.770213   775.852660    1326.571862  
BLD 08

In [19]:
df[(df['Meter Scope'] == 'BLD 1 - 9')].groupby('Location').mean()

TDS #    EDP  UMIS BILL ID     # days  Current Charges  \
Location                                                           
BLD 02     55.0  525.0  3.781606e+06  27.008929     44982.836339   

          Consumption (KWH)   KWH Charges  Consumption (KW)   KW Charges  \
Location                                                                   
BLD 02        290678.571429  15647.757054        475.214286  4529.356607   

          Other charges  
Location                 
BLD 02     22893.834732

### 6. Create an unique identifier for each building and remove unnecessary fields

In [20]:
# the combination of TDS# and Location uniquely determines a buildling
# Use EDP or RC Code when TDS# is not available
df['Building ID'] = df['TDS #'].combine_first(df['EDP']).map(str).combine_first(df['RC Code']) \
                    + " - " + df['Location'].map(lambda x: 'NA' if pd.isna(x) else x)

In [21]:
# Building ID alone is not the primary key of the data
df.groupby(['Building ID', 'Revenue Month']).count().shape[0]/df.shape[0]

0.6323396567299007

In [22]:
# the combination of Building ID, meter number and revenue month is still not a primary key
df.groupby(['Building ID', 'Meter Number', 'Revenue Month']).count().shape[0]/df.shape[0]

0.9987636458285487

In [23]:
# Define a list of columns of interest
cols = ['Account Name', 'Location', 'Building ID', 'Meter Number',
        'Revenue Month', 'Service Start Date', 'Service End Date', '# days', 
       'Current Charges','Consumption (KWH)', 'KWH Charges',
       'Consumption (KW)', 'KW Charges', 'Other charges']
df = df[cols]

#### In some cases multiple rows exists for one revenue month, with consecutive service start and end dates. We need to combined these rows for the same meter-month

In [24]:
df[(df['Building ID'] == '98.0 - BLD 15') & (df['Meter Number'] == 99277502) & (df['Revenue Month'] == '2013-03-01')][cols]

Account Name Location    Building ID Meter Number  \
117025  OCEAN BAY APARTMENTS (BAYSIDE)   BLD 15  98.0 - BLD 15     99277502   
117026  OCEAN BAY APARTMENTS (BAYSIDE)   BLD 15  98.0 - BLD 15     99277502   
117027  OCEAN BAY APARTMENTS (BAYSIDE)   BLD 15  98.0 - BLD 15     99277502   

       Revenue Month Service Start Date Service End Date  # days  \
117025    2013-03-01         2013-02-19       2013-03-18    27.0   
117026    2013-03-01         2013-03-18       2013-03-19     1.0   
117027    2013-03-01         2013-03-19       2013-03-26     7.0   

        Current Charges  Consumption (KWH)  KWH Charges  Consumption (KW)  \
117025         62086.60           398400.0     40709.83             838.8   
117026            14.64                0.0         0.00               0.0   
117027           103.84                0.0         0.00               1.2   

        KW Charges  Other charges  
117025         0.0       21376.77  
117026         0.0          14.64  
117027         0.0         103.84

In [25]:
df[(df['Building ID'] == '165.0 - BLD 01') & (df['Meter Number'] == '099800299') & (df['Revenue Month'] == '2016-06-01')][cols]

Empty DataFrame
Columns: [Account Name, Location, Building ID, Meter Number, Revenue Month, Service Start Date, Service End Date, # days, Current Charges, Consumption (KWH), KWH Charges, Consumption (KW), KW Charges, Other charges]
Index: []

In [26]:
# the combination of Building ID, meter number and revenue month is almost a primary key
df.groupby(['Building ID', 'Meter Number', 'Revenue Month', 'Service Start Date', 'Service End Date']).count().shape[0]/df.shape[0]

0.9994870852663329

### 7. There are duplicated rows in the dataset - remove duplicates in df

In [27]:
df = df.drop_duplicates()

### 8. Check which combinations of the 5 fields (Building ID, Meter, Month, StartDate, EndDate) has multiple rows and why

In [28]:
idx = df.groupby(['Building ID', 'Meter Number', 'Revenue Month', 'Service Start Date', 'Service End Date']).count()['Account Name'].reset_index()
idx.columns = ['Building ID', 'Meter Number', 'Revenue Month','Service Start Date', 'Service End Date', 'Counts']
idx = idx[idx['Counts'] > 1]

dupRows = idx.sort_values('Counts', ascending = False)

a = pd.merge(dupRows.iloc[:, 0:3], df[cols], on = \
         ['Building ID', 'Meter Number', 'Revenue Month'], how = 'inner')[cols]\
        .sort_values(['Building ID', 'Meter Number', 'Revenue Month'])

#### half of these problematic rows has zero values in the numerical fields like "current charges"

In [29]:
a

Account Name           Location               Building ID  \
0             KINGSBOROUGH             BLD 06             10.0 - BLD 06   
1             KINGSBOROUGH             BLD 06             10.0 - BLD 06   
36            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
37            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
40            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
41            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
42            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
43            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
44            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
45            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
46            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
47            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
48            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
49            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
50            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
51            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
38            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
39            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
52            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
53            KINGSBOROUGH             BLD 06             10.0 - BLD 06   
56                 JACKSON                NaN                120.0 - NA   
57                 JACKSON                NaN                120.0 - NA   
58              LOW HOUSES             BLD 04            169.0 - BLD 04   
59              LOW HOUSES             BLD 04            169.0 - BLD 04   
60              LOW HOUSES             BLD 04            169.0 - BLD 04   
61              LOW HOUSES             BLD 04            169.0 - BLD 04   
62              LOW HOUSES             BLD 04            169.0 - BLD 04   
63              LOW HOUSES             BLD 04            169.0 - BLD 04   
64              LOW HOUSES             BLD 04            169.0 - BLD 04   
65              LOW HOUSES             BLD 04            169.0 - BLD 04   
..                     ...                ...                       ...   
4             THROGGS NECK             BLD 11             63.0 - BLD 11   
5             THROGGS NECK             BLD 11             63.0 - BLD 11   
6             THROGGS NECK             BLD 14             63.0 - BLD 14   
7             THROGGS NECK             BLD 14             63.0 - BLD 14   
8             THROGGS NECK             BLD 14             63.0 - BLD 14   
9             THROGGS NECK             BLD 14             63.0 - BLD 14   
10            THROGGS NECK             BLD 14             63.0 - BLD 14   
11            THROGGS NECK             BLD 14             63.0 - BLD 14   
12            THROGGS NECK             BLD 14             63.0 - BLD 14   
13            THROGGS NECK             BLD 14             63.0 - BLD 14   
14            THROGGS NECK             BLD 14             63.0 - BLD 14   
15            THROGGS NECK             BLD 14             63.0 - BLD 14   
2             THROGGS NECK             BLD 14             63.0 - BLD 14   
3             THROGGS NECK             BLD 14             63.0 - BLD 14   
16            THROGGS NECK             BLD 14             63.0 - BLD 14   
17            THROGGS NECK             BLD 14             63.0 - BLD 14   
20            THROGGS NECK             BLD 14             63.0 - BLD 14   
21            THROGGS NECK             BLD 14             63.0 - BLD 14   
22  DOUGLASS I/DOUGLASS II  DOUGLASS I BLD 05  82.0 - DOUGLASS I BLD 05   
23  DOUGLASS I/DOUGLASS II  DOUGLASS I BLD 05  82.0 - DOUGLASS I BLD 05   
24  DOUGLASS I/DOUGLASS II  DOUGLASS I BLD 06  82.0 - DOUGLASS I BLD 06   
25  DOUGLASS I/DOUGLASS II  DOUGLASS I 

#### remove those rows from the dataset

In [30]:
df = df[~((df['Current Charges'] == 0) & (df['KWH Charges'] == 0) & (df['KW Charges'] == 0) \
  & (df['Other charges'] == 0) & (df['Consumption (KWH)'] == 0) & (df['Consumption (KW)'] == 0))]

In [31]:
idx = df.groupby(['Building ID', 'Meter Number', 'Revenue Month', 'Service Start Date', 'Service End Date']).count()['Account Name'].reset_index()
idx.columns = ['Building ID', 'Meter Number', 'Revenue Month','Service Start Date', 'Service End Date', 'Counts']
idx = idx[idx['Counts'] > 1]

dupRows = idx.sort_values('Counts', ascending = False)

a = pd.merge(dupRows.iloc[:, 0:3], df[cols], on = \
         ['Building ID', 'Meter Number', 'Revenue Month'], how = 'inner')[cols]\
        .sort_values(['Building ID', 'Meter Number', 'Revenue Month'])

#### Only one row left, seems a case of rebilling

In [32]:
a

Account Name Location    Building ID Meter Number Revenue Month  \
0  THROGGS NECK   BLD 11  63.0 - BLD 11      8125318    2011-10-01   
1  THROGGS NECK   BLD 11  63.0 - BLD 11      8125318    2011-10-01   

  Service Start Date Service End Date  # days  Current Charges  \
0         2011-09-22       2011-10-24    32.0          1306.02   
1         2011-09-22       2011-10-24    32.0          2693.18   

   Consumption (KWH)  KWH Charges  Consumption (KW)  KW Charges  Other charges  
0            12880.0       858.84               0.0         0.0         447.18  
1            26560.0      1771.02               0.0         0.0         922.16

### 9. Combine the rows of same Building, Meter and Month

In [33]:
import numpy as np

In [34]:
dfCharges = pd.pivot_table(df, values = ['# days', 'Current Charges','Consumption (KWH)', 'KWH Charges',\
       'Consumption (KW)', 'KW Charges', 'Other charges'], index=['Account Name', 'Location', 'Building ID', 'Meter Number',
       'Revenue Month'], aggfunc = np.sum).reset_index()

dfStartDate = pd.pivot_table(df, values = ['Service Start Date'], index=['Account Name', 'Location', 'Building ID', 'Meter Number',
       'Revenue Month'], aggfunc = np.min).reset_index()

dfEndDate = pd.pivot_table(df, values = ['Service End Date'], index=['Account Name', 'Location', 'Building ID', 'Meter Number',
       'Revenue Month'], aggfunc = np.max).reset_index()

In [35]:
df_combined = pd.merge(dfStartDate, dfEndDate, on = ['Account Name', 'Location', 'Building ID', 'Meter Number',
       'Revenue Month'], how = 'inner')

df_combined = pd.merge(df_combined, dfCharges, on = ['Account Name', 'Location', 'Building ID', 'Meter Number',
       'Revenue Month'], how = 'inner')

In [36]:
df_combined.head()

Account Name Location     Building ID Meter Number Revenue Month  \
0        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-01-01   
1        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-02-01   
2        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-03-01   
3        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-04-01   
4        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-05-01   

  Service Start Date Service End Date  # days  Consumption (KW)  \
0         2009-12-24       2010-01-26    33.0             148.0   
1         2010-01-26       2010-02-25    30.0             144.0   
2         2010-02-25       2010-03-26    29.0             136.0   
3         2010-03-26       2010-04-26    31.0             124.0   
4         2010-04-26       2010-05-24    28.0             144.0   

   Consumption (KWH)  Current Charges  KW Charges  KWH Charges  Other charges  
0            84000.0         10314.51      1924.0      4818.24        3572.27  
1            75200.0          9422.06      1872.0      4313.47        3236.59  
2            68800.0          8988.12      1768.0      3946.37        3273.75  
3            68400.0          9146.17      1612.0      3923.42        3610.75  
4            64800.0          9137.42      1872.0      3716.93        3548.49

### 10. save a copy of the original dataframe

In [37]:
df_orig = df
df = df_combined

### 11. Add new metrics

In [38]:
df['Total Charges'] = df['KW Charges'] + df['KWH Charges']

In [39]:
df['Total Energy Rate'] = df['Total Charges']/df['Consumption (KWH)']

In [40]:
df['Total Charge Match'] = (df['Current Charges'] == df['KWH Charges'] + df['KW Charges'] + df['Other charges'])

In [41]:
df.head()

Account Name Location     Building ID Meter Number Revenue Month  \
0        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-01-01   
1        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-02-01   
2        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-03-01   
3        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-04-01   
4        ADAMS   BLD 01  118.0 - BLD 01      7518352    2010-05-01   

  Service Start Date Service End Date  # days  Consumption (KW)  \
0         2009-12-24       2010-01-26    33.0             148.0   
1         2010-01-26       2010-02-25    30.0             144.0   
2         2010-02-25       2010-03-26    29.0             136.0   
3         2010-03-26       2010-04-26    31.0             124.0   
4         2010-04-26       2010-05-24    28.0             144.0   

   Consumption (KWH)  Current Charges  KW Charges  KWH Charges  Other charges  \
0            84000.0         10314.51      1924.0      4818.24        3572.27   
1            75200.0          9422.06      1872.0      4313.47        3236.59   
2            68800.0          8988.12      1768.0      3946.37        3273.75   
3            68400.0          9146.17      1612.0      3923.42        3610.75   
4            64800.0          9137.42      1872.0      3716.93        3548.49   

   Total Charges  Total Energy Rate  Total Charge Match  
0        6742.24           0.080265                True  
1        6185.47           0.082254               False  
2        5714.37           0.083058               False  
3        5535.42           0.080927                True  
4        5588.93           0.086249                True

### 12. Save the cleaned data to the output folder

In [47]:
df.to_csv('../output/NYCHA_Electricity_2010_to_2018_Cleaned.csv')

### 13. Investigate the zero values in the fields of charges and consumption (Incomplete)

#### check how much % of rows have zero values in each numeric field

In [48]:
df[(df['KWH Charges'] == 0) ].shape[0]\
    /df.shape[0]

0.3268528710691575

In [49]:
df[(df['Consumption (KWH)'] == 0)  ].shape[0]\
    /df.shape[0]

0.3244731819729642

In [50]:
df[(df['KW Charges'] == 0) ].shape[0]\
    /df.shape[0]

0.41077155776586394

In [51]:
df[ (df['Consumption (KW)'] == 0) ].shape[0]\
    /df.shape[0]

0.3702495307934145

In [52]:
df[ (df['Current Charges'] == 0) ].shape[0]\
    /df.shape[0]

0.16495878109236045

In [53]:
df[ (df['Other charges'] == 0) ].shape[0]\
    /df.shape[0]

0.11739667556245396

#### check the correlation of numerical fields having zero values

In [54]:
df['KWH_Charge_0'] = df['KWH Charges'] == 0
df['KWH_0'] = df['Consumption (KWH)'] == 0
df['KW_Charge_0'] = df['KW Charges'] == 0
df['KW_0'] = df['Consumption (KW)'] == 0
df['Other_Charge_0'] = df['Other charges'] == 0

In [55]:
df[(df['KWH Charges'] == 0) | (df['Consumption (KWH)'] == 0) \
   | (df['KW Charges'] == 0) | (df['Consumption (KW)'] == 0) ].shape[0]\
    /df.shape[0]

0.7342825692723141

In [56]:
df[((df['KWH Charges'] == 0) ^ (df['Consumption (KWH)'] == 0)) \
   | ((df['KW Charges'] == 0) ^ (df['Consumption (KW)'] == 0)) ].shape[0]\
    /df.shape[0]

0.045190335532203016

In [57]:
df[((df['KWH Charges'] == 0) & (df['Consumption (KWH)'] == 0)) ]\
.groupby(['KWH_Charge_0', 'KWH_0', 'KW_Charge_0', 'KW_0']).agg(['count'])

Account Name Location Building ID  \
                                            count    count       count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                      
True         True  False       False        81375    81375       81375   
                               True            12       12          12   
                   True        False          103      103         103   
                               True           447      447         447   

                                     Meter Number Revenue Month  \
                                            count         count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                               
True         True  False       False        81375         81375   
                               True            12            12   
                   True        False          103           103   
                               True           447           447   

                                     Service Start Date Service End Date  \
                                                  count            count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                        
True         True  False       False              81375            81375   
                               True                  12               12   
                   True        False                103              103   
                               True                 447              447   

                                     # days Consumption (KW)  \
                                      count            count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                            
True         True  False       False  81375            81375   
                               True      12               12   
                   True        False    103              103   
                               True     447              447   

                                     Consumption (KWH) Current Charges  \
                                                 count           count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                      
True         True  False       False             81375           81375   
                               True                 12              12   
                   True        False               103             103   
                               True                447             447   

                                     KW Charges KWH Charges Other charges  \
                                          count       count         count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                         
True         True  False       False      81375       81375         81375   
                               True          12          12            12   
                   True        False        103         103           103   
                               True         447         447           447   

                                     Total Charges Total Energy Rate  \
                                             count             count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                    
True         True  False       False         81375             81375   
                               True             12                12   
                   True        False           103                 0   
                               True            447                 0   

                                     Total Charge Match Other_Charge_0  
                                                  count          count  
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                     
True         True  False       False              81375          81375  
                               True                  12             12  
                   True        False                103            103  
                               True                 447         

In [58]:
df[((df['KWH Charges'] == 0) ^ (df['Consumption (KWH)'] == 0)) ]\
.groupby(['KWH_Charge_0', 'KWH_0', 'KW_Charge_0', 'KW_0']).agg(['count'])

Account Name Location Building ID  \
                                            count    count       count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                      
False        True  False       True             7        7           7   
                   True        True             3        3           3   
True         False False       False           17       17          17   
                   True        False           59       59          59   
                               True           535      535         535   

                                     Meter Number Revenue Month  \
                                            count         count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                               
False        True  False       True             7             7   
                   True        True             3             3   
True         False False       False           17            17   
                   True        False           59            59   
                               True           535           535   

                                     Service Start Date Service End Date  \
                                                  count            count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                        
False        True  False       True                   7                7   
                   True        True                   3                3   
True         False False       False                 17               17   
                   True        False                 59               59   
                               True                 535              535   

                                     # days Consumption (KW)  \
                                      count            count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                            
False        True  False       True       7                7   
                   True        True       3                3   
True         False False       False     17               17   
                   True        False     59               59   
                               True     535              535   

                                     Consumption (KWH) Current Charges  \
                                                 count           count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                      
False        True  False       True                  7               7   
                   True        True                  3               3   
True         False False       False                17              17   
                   True        False                59              59   
                               True                535             535   

                                     KW Charges KWH Charges Other charges  \
                                          count       count         count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                         
False        True  False       True           7           7             7   
                   True        True           3           3             3   
True         False False       False         17          17            17   
                   True        False         59          59            59   
                               True         535         535           535   

                                     Total Charges Total Energy Rate  \
                                             count             count   
KWH_Charge_0 KWH_0 KW_Charge_0 KW_0                                    
False        True  False       True              7                 7   
                   True        True              3                 3   
True         False False       False            17                17   
                   True        False            59                59   
                               True            535               535   

In [59]:
df[((df['KW Charges'] == 0) & (df['Consumption (KW)'] == 0))]\
.groupby(['KW_Charge_0', 'KW_0', 'KWH_Charge_0', 'KWH_0']).agg(['count'])

Account Name Location Building ID  \
                                           count    count       count   
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                                     
True        True False        False        92211    92211       92211   
                              True             3        3           3   
                 True         False          535      535         535   
                              True           447      447         447   

                                    Meter Number Revenue Month  \
                                           count         count   
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                              
True        True False        False        92211         92211   
                              True             3             3   
                 True         False          535           535   
                              True           447           447   

                                    Service Start Date Service End Date  \
                                                 count            count   
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                                       
True        True False        False              92211            92211   
                              True                   3                3   
                 True         False                535              535   
                              True                 447              447   

                                    # days Consumption (KW) Consumption (KWH)  \
                                     count            count             count   
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                                             
True        True False        False  92211            92211             92211   
                              True       3                3                 3   
                 True         False    535              535               535   
                              True     447              447               447   

                                    Current Charges KW Charges KWH Charges  \
                                              count      count       count   
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                                          
True        True False        False           92211      92211       92211   
                              True                3          3           3   
                 True         False             535        535         535   
                              True              447        447         447   

                                    Other charges Total Charges  \
                                            count         count   
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                               
True        True False        False         92211         92211   
                              True              3             3   
                 True         False           535           535   
                              True            447           447   

                                    Total Energy Rate Total Charge Match  \
                                                count              count   
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                                        
True        True False        False             92211              92211   
                              True                  3                  3   
                 True         False               535                535   
                              True                  0                447   

                                    Other_Charge_0  
                                             count  
KW_Charge_0 KW_0 KWH_Charge_0 KWH_0                 
True        True False        False          92211  
                              True               3  
                 True         False            535  
                              True             447

In [60]:
df[((df['KW Charges'] == 0) ^ (df['Consumption (KW)'] == 0))]\
.groupby(['KW_Charge_0', 'KW_0', 'KWH_Charge_0', 'KWH_0']).agg(['count'])

Account Name Location Building ID  \
                                            count    count       count   
KW_Charge_0 KW_0  KWH_Charge_0 KWH_0                                     
False       True  False        False          293      293         293   
                               True             7        7           7   
                  True         True            12       12          12   
True        False False        False        10384    10384       10384   
                  True         False           59       59          59   
                               True           103      103         103   

                                     Meter Number Revenue Month  \
                                            count         count   
KW_Charge_0 KW_0  KWH_Charge_0 KWH_0                              
False       True  False        False          293           293   
                               True             7             7   
                  True         True            12            12   
True        False False        False        10384         10384   
                  True         False           59            59   
                               True           103           103   

                                     Service Start Date Service End Date  \
                                                  count            count   
KW_Charge_0 KW_0  KWH_Charge_0 KWH_0                                       
False       True  False        False                293              293   
                               True                   7                7   
                  True         True                  12               12   
True        False False        False              10384            10384   
                  True         False                 59               59   
                               True                 103              103   

                                     # days Consumption (KW)  \
                                      count            count   
KW_Charge_0 KW_0  KWH_Charge_0 KWH_0                           
False       True  False        False    293              293   
                               True       7                7   
                  True         True      12               12   
True        False False        False  10384            10384   
                  True         False     59               59   
                               True     103              103   

                                     Consumption (KWH) Current Charges  \
                                                 count           count   
KW_Charge_0 KW_0  KWH_Charge_0 KWH_0                                     
False       True  False        False               293             293   
                               True                  7               7   
                  True         True                 12              12   
True        False False        False             10384           10384   
                  True         False                59              59   
                               True                103             103   

                                     KW Charges KWH Charges Other charges  \
                                          count       count         count   
KW_Charge_0 KW_0  KWH_Charge_0 KWH_0                                        
False       True  False        False        293         293           293   
                               True           7           7             7   
                  True         True          12          12            12   
True        False False        False      10384       10384         10384   
                  True         False         59          59            59   
                               True         103         103           103   

                                     Total Charges Total Energy Rate  \
                                             count             count   

## Q&A with Linnea:

1. why would "Consumption (KW)" be zero?
    KW and KWH should be both positive, unless there are some related bills that already covers it
2. What's the "Other Charges"?
    - negative values to adjust for the payments from previous month
    - taxes, fee for meter-reading, little fees charged by utilities and states (e.g. system benefit charge), credit (state got a better deal after charging the clients)

## To Do:

1. investigate the zero values in the fields of charges and consumption
    - current charges = 0 and other charges != 0
    - current charges == other charges and current charges != 0
2. investigate why (Current Charge) != (KWH charges) + (KW charges) + (Other Charges)
3. check if for the same building, meter number varies across different months
4. check rebills (can probably detected by variance-checking)
5. check gaps in service and whether the union of all the billing window is one-year for each meter
    - start, end dates are useful when there is not a complete month reported (1-3 day gap is okay , 5 days is not)
6. anomaly detection of charges & consumptions at the meter level and building level

**Issue- 2 - Checking for cases where current charge is not equal to the sum of KWH, kW and Other charges**

In [3]:
import pandas as pd
clean_df = pd.read_csv("NYCHA_Electricity_2010_to_2018_Cleaned.csv", index_col=False)

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
# create a new column that aggregates all charge values
clean_df["sum_all_charges"] = clean_df.apply(lambda x: x['KW Charges'] + x['KWH Charges'] + x['Other charges'], axis=1)

In [5]:
# check if the current charges are equal to the sum of all charges column
clean_df["check_current_charge"] = clean_df.apply(lambda x: round(x['Current Charges']) == round(x['sum_all_charges']), axis=1)

In [6]:
# the number of rows with mismatching values
clean_df['check_current_charge'].value_counts()

True     217503
False     35051
Name: check_current_charge, dtype: int64

In [7]:
# percentage of rows with mismatching values in the clean data
(len(clean_df[(clean_df.check_current_charge == False)])/len(clean_df['check_current_charge']))*100

13.878616058347918

In [8]:
clean_df[(clean_df.check_current_charge == False)]

Account Name Location     Building ID Meter Number Revenue Month  \
55                ADAMS   BLD 01  118.0 - BLD 01      7518352        1/1/15   
56                ADAMS   BLD 01  118.0 - BLD 01      7518352        2/1/15   
57                ADAMS   BLD 01  118.0 - BLD 01      7518352        3/1/15   
58                ADAMS   BLD 01  118.0 - BLD 01      7518352        4/1/15   
59                ADAMS   BLD 01  118.0 - BLD 01      7518352        5/1/15   
60                ADAMS   BLD 01  118.0 - BLD 01      7518352        6/1/15   
61                ADAMS   BLD 01  118.0 - BLD 01      7518352        8/1/15   
62                ADAMS   BLD 01  118.0 - BLD 01      7518352        9/1/15   
63                ADAMS   BLD 01  118.0 - BLD 01      7518352       10/1/15   
64                ADAMS   BLD 01  118.0 - BLD 01      7518352       11/1/15   
145               ADAMS   BLD 02  118.0 - BLD 02      7725456        1/1/15   
146               ADAMS   BLD 02  118.0 - BLD 02      7725456        2/1/15   
147               ADAMS   BLD 02  118.0 - BLD 02      7725456        3/1/15   
148               ADAMS   BLD 02  118.0 - BLD 02      7725456        4/1/15   
149               ADAMS   BLD 02  118.0 - BLD 02      7725456        5/1/15   
150               ADAMS   BLD 02  118.0 - BLD 02      7725456        6/1/15   
151               ADAMS   BLD 02  118.0 - BLD 02      7725456        7/1/15   
152               ADAMS   BLD 02  118.0 - BLD 02      7725456        8/1/15   
153               ADAMS   BLD 02  118.0 - BLD 02      7725456        9/1/15   
154               ADAMS   BLD 02  118.0 - BLD 02      7725456       10/1/15   
155               ADAMS   BLD 02  118.0 - BLD 02      7725456       11/1/15   
238               ADAMS   BLD 03  118.0 - BLD 03      7725454        1/1/15   
239               ADAMS   BLD 03  118.0 - BLD 03      7725454        2/1/15   
240               ADAMS   BLD 03  118.0 - BLD 03      7725454        3/1/15   
241               ADAMS   BLD 03  118.0 - BLD 03      7725454        4/1/15   
242               ADAMS   BLD 03  118.0 - BLD 03      7725454        5/1/15   
243               ADAMS   BLD 03  118.0 - BLD 03      7725454        6/1/15   
244               ADAMS   BLD 03  118.0 - BLD 03      7725454        7/1/15   
245               ADAMS   BLD 03  118.0 - BLD 03      7725454        8/1/15   
246               ADAMS   BLD 03  118.0 - BLD 03      7725454        9/1/15   
...                 ...      ...             ...          ...           ...   
252314  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      1096667        7/1/15   
252315  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      1096667        8/1/15   
252316  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      1096667        9/1/15   
252317  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      1096667       10/1/15   
252318  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      1096667       11/1/15   
252319  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      1096667       12/1/15   
252410  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        1/1/15   
252411  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        2/1/15   
252412  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        3/1/15   
252413  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        4/1/15   
252414  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        5/1/15   
252415  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        6/1/15   
252416  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        7/1/15   
252417  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        8/1/15   
252418  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664        9/1/15   
252419  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664       10/1/15   
252420  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664       11/1/15   
252421  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096664       12/1/15   
252512  WYCKOFF GARDENS   BLD 03  163.0 - BLD 03      8096667

**Looking at the data for which the values are not the same, nothing out of the ordinary seems to stand out. It could be due to missing fields, typos while entering the data, or unknown parameters, even a combination of all those factors.**

**Issue - 3 - Aggregating meter numbers for each building id, to identify buildings with multiple meter readings**

In [9]:
# list of all buildings with number of meters used to collect data
clean_df.groupby('Building ID').agg({'Meter Number': 'nunique'})

Meter Number
Building ID                                    
1.0 - BLD 01                                  2
1.0 - BLD 02                                  1
1.0 - BLD 03                                  1
1.0 - BLD 04                                  3
1.0 - BLD 05                                  1
1.0 - BLD 06                                  1
1.0 - BLD 07                                  2
10.0 - BLD 01                                 2
10.0 - BLD 02                                 3
10.0 - BLD 03                                 2
10.0 - BLD 04                                 2
10.0 - BLD 05                                 4
10.0 - BLD 06                                 4
10.0 - BLD 07                                 2
10.0 - BLD 08                                 2
10.0 - BLD 09                                 4
10.0 - BLD 10                                 2
10.0 - BLD 11                                 3
10.0 - BLD 13                                 4
10.0 - BLD 14                                 4
10.0 - BLD 15                                 2
10.0 - BLD 16                                 7
100.0 - BLD 01                                1
100.0 - BLD 02                                1
100.0 - BLD 03 - Community Center             2
101.0 - BLD 02                                2
101.0 - BLD 04                                2
102.0 - BLD 02                                2
102.0 - BLD 07                                2
102.0 - BLD 10                                2
...                                         ...
95.0 - LINDEN BLD 20                          3
96.0 - BLD 01                                 1
96.0 - BLD 02                                 2
96.0 - BLD 03                                 1
96.0 - BLD 04                                 1
96.0 - BLD 05                                 1
96.0 - BLD 06                                 1
96.0 - BLD 07                                 1
96.0 - BLD 08                                 1
97.0 - BLD 01                                 1
97.0 - BLD 02                                 1
97.0 - BLD 03                                 2
97.0 - BLD 04                                 1
97.0 - BLD 05                                 1
97.0 - BLD 09                                 3
97.0 - BLD 10 - Community Center              3
98.0 - BLD 14                                 4
98.0 - BLD 15                                 2
99.0 - BLD 01                                 1
99.0 - BLD 02                                 1
99.0 - BLD 03                                 1
99.0 - BLD 04                                 1
99.0 - BLD 05                                 1
999.0 - BLD 01                               23
999.0 - BLD 03                                3
999.0 - BLD 04                                1
999.0 - BLD 05                                1
999.0 - BLD 06                                2
999.0 - BLD 07                                2
999.0 - BLD 08                                1

[1994 rows x 1 columns]

In [10]:
# filtering those rows for which count > 1
temp = clean_df.groupby('Building ID').agg({'Meter Number': 'nunique'})
temp = temp.apply(lambda x: x['Meter Number'] > 1, axis = 1)
temp = temp.to_frame()
temp.columns = ['multiple_meters']
temp.query('multiple_meters')

multiple_meters
Building ID                                        
1.0 - BLD 01                                   True
1.0 - BLD 04                                   True
1.0 - BLD 07                                   True
10.0 - BLD 01                                  True
10.0 - BLD 02                                  True
10.0 - BLD 03                                  True
10.0 - BLD 04                                  True
10.0 - BLD 05                                  True
10.0 - BLD 06                                  True
10.0 - BLD 07                                  True
10.0 - BLD 08                                  True
10.0 - BLD 09                                  True
10.0 - BLD 10                                  True
10.0 - BLD 11                                  True
10.0 - BLD 13                                  True
10.0 - BLD 14                                  True
10.0 - BLD 15                                  True
10.0 - BLD 16                                  True
100.0 - BLD 03 - Community Center              True
101.0 - BLD 02                                 True
101.0 - BLD 04                                 True
102.0 - BLD 02                                 True
102.0 - BLD 07                                 True
102.0 - BLD 10                                 True
102.0 - MORRIS I BLD 05                        True
103.0 - BLD 04                                 True
103.0 - BLD 05                                 True
11.0 - CLASON POINT GARDENS BLD 03             True
11.0 - CLASON POINT GARDENS BLD 13             True
11.0 - CLASON POINT GARDENS BLD 26             True
...                                             ...
94.0 - BLD 05                                  True
95.0 - LINDEN BLD 01                           True
95.0 - LINDEN BLD 02                           True
95.0 - LINDEN BLD 03                           True
95.0 - LINDEN BLD 04                           True
95.0 - LINDEN BLD 05                           True
95.0 - LINDEN BLD 06                           True
95.0 - LINDEN BLD 07                           True
95.0 - LINDEN BLD 08                           True
95.0 - LINDEN BLD 10                           True
95.0 - LINDEN BLD 11                           True
95.0 - LINDEN BLD 12                           True
95.0 - LINDEN BLD 13                           True
95.0 - LINDEN BLD 14                           True
95.0 - LINDEN BLD 15                           True
95.0 - LINDEN BLD 16                           True
95.0 - LINDEN BLD 17                           True
95.0 - LINDEN BLD 18                           True
95.0 - LINDEN BLD 19                           True
95.0 - LINDEN BLD 20                           True
96.0 - BLD 02                                  True
97.0 - BLD 03                                  True
97.0 - BLD 09                                  True
97.0 - BLD 10 - Community Center               True
98.0 - BLD 14                                  True
98.0 - BLD 15                                  True
999.0 - BLD 01                                 True
999.0 - BLD 03                                 True
999.0 - BLD 06                                 True
999.0 - BLD 07                                 True

[1396 rows x 1 columns]

**A large number of buildings (1396 rows) seem to be using multiple meters to collect data. Need to investigate if this causes any overlap in the stored data**

**Issue-5, checking if the service window covers the entire month, for each building id**

In [11]:
clean_df[clean_df['Building ID'] == '118.0 - BLD 01']

Account Name Location     Building ID Meter Number Revenue Month  \
0         ADAMS   BLD 01  118.0 - BLD 01      7518352        1/1/10   
1         ADAMS   BLD 01  118.0 - BLD 01      7518352        2/1/10   
2         ADAMS   BLD 01  118.0 - BLD 01      7518352        3/1/10   
3         ADAMS   BLD 01  118.0 - BLD 01      7518352        4/1/10   
4         ADAMS   BLD 01  118.0 - BLD 01      7518352        5/1/10   
5         ADAMS   BLD 01  118.0 - BLD 01      7518352        6/1/10   
6         ADAMS   BLD 01  118.0 - BLD 01      7518352        7/1/10   
7         ADAMS   BLD 01  118.0 - BLD 01      7518352        9/1/10   
8         ADAMS   BLD 01  118.0 - BLD 01      7518352       10/1/10   
9         ADAMS   BLD 01  118.0 - BLD 01      7518352       11/1/10   
10        ADAMS   BLD 01  118.0 - BLD 01      7518352       12/1/10   
11        ADAMS   BLD 01  118.0 - BLD 01      7518352        3/1/11   
12        ADAMS   BLD 01  118.0 - BLD 01      7518352        4/1/11   
13        ADAMS   BLD 01  118.0 - BLD 01      7518352        5/1/11   
14        ADAMS   BLD 01  118.0 - BLD 01      7518352        6/1/11   
15        ADAMS   BLD 01  118.0 - BLD 01      7518352        8/1/11   
16        ADAMS   BLD 01  118.0 - BLD 01      7518352        9/1/11   
17        ADAMS   BLD 01  118.0 - BLD 01      7518352       10/1/11   
18        ADAMS   BLD 01  118.0 - BLD 01      7518352       11/1/11   
19        ADAMS   BLD 01  118.0 - BLD 01      7518352       12/1/11   
20        ADAMS   BLD 01  118.0 - BLD 01      7518352        1/1/12   
21        ADAMS   BLD 01  118.0 - BLD 01      7518352        2/1/12   
22        ADAMS   BLD 01  118.0 - BLD 01      7518352        3/1/12   
23        ADAMS   BLD 01  118.0 - BLD 01      7518352        4/1/12   
24        ADAMS   BLD 01  118.0 - BLD 01      7518352        5/1/12   
25        ADAMS   BLD 01  118.0 - BLD 01      7518352        6/1/12   
26        ADAMS   BLD 01  118.0 - BLD 01      7518352        7/1/12   
27        ADAMS   BLD 01  118.0 - BLD 01      7518352        8/1/12   
28        ADAMS   BLD 01  118.0 - BLD 01      7518352        9/1/12   
29        ADAMS   BLD 01  118.0 - BLD 01      7518352       11/1/12   
..          ...      ...             ...          ...           ...   
59        ADAMS   BLD 01  118.0 - BLD 01      7518352        5/1/15   
60        ADAMS   BLD 01  118.0 - BLD 01      7518352        6/1/15   
61        ADAMS   BLD 01  118.0 - BLD 01      7518352        8/1/15   
62        ADAMS   BLD 01  118.0 - BLD 01      7518352        9/1/15   
63        ADAMS   BLD 01  118.0 - BLD 01      7518352       10/1/15   
64        ADAMS   BLD 01  118.0 - BLD 01      7518352       11/1/15   
65        ADAMS   BLD 01  118.0 - BLD 01      7518352        4/1/16   
66        ADAMS   BLD 01  118.0 - BLD 01      7518352        5/1/16   
67        ADAMS   BLD 01  118.0 - BLD 01      7518352        6/1/16   
68        ADAMS   BLD 01  118.0 - BLD 01      7518352        7/1/16   
69        ADAMS   BLD 01  118.0 - BLD 01      7518352        8/1/16   
70        ADAMS   BLD 01  118.0 - BLD 01      7518352        9/1/16   
71        ADAMS   BLD 01  118.0 - BLD 01      7518352       10/1/16   
72        ADAMS   BLD 01  118.0 - BLD 01      7518352       11/1/16   
73        ADAMS   BLD 01  118.0 - BLD 01      7518352       12/1/16   
74        ADAMS   BLD 01  118.0 - BLD 01      7518352        1/1/17   
75        ADAMS   BLD 01  118.0 - BLD 01      7518352        2/1/17   
76        ADAMS   BLD 01  118.0 - BLD 01      7518352        3/1/17   
77        ADAMS   BLD 01  118.0 - BLD 01      7518352        4/1/17   
78        ADAMS   BLD 01  118.0 - BLD 01      7518352        6/1/17   
79        ADAMS   BLD 01  118.0 - BLD 01      7518352        7/1/17   
80        ADAMS   BLD 01  118.0 - BLD 01      7518352        8/1/17   
81        ADAMS   BLD 01  118.0 - BLD 01      7518352        9/1/17   
82        ADAMS   BLD 01  118.0 - BLD 01      7518352       10/1/17   
83        ADAMS   BLD 01  1

In [12]:
# create a new column for revenue year
clean_df['Revenue Month'] = pd.to_datetime(clean_df['Revenue Month'])
clean_df['Service Start Date'] = pd.to_datetime(clean_df['Service Start Date'])
clean_df['Service End Date'] = pd.to_datetime(clean_df['Service End Date'])
clean_df['revenue_year'] = clean_df['Revenue Month'].dt.year


In [16]:
# group by building id, and revenue year, and find sum of # days in each year

clean_df.groupby(['Building ID','revenue_year']).agg({'# days': sum})

# days
Building ID    revenue_year        
1.0 - BLD 01   2010             639
               2011             551
               2012             670
               2013             730
               2014             666
               2015             608
               2016             302
               2017             302
               2018             364
1.0 - BLD 02   2010             364
               2011             335
               2012             335
               2013             365
               2014             333
               2015             304
               2016             151
               2017             151
               2018             182
1.0 - BLD 03   2010             364
               2011             335
               2012             335
               2013             365
               2014             333
               2015             304
               2016             151
               2017             123
               2018             182
1.0 - BLD 04   2010             364
               2011             335
               2012             335
...                             ...
999.0 - BLD 03 2010             666
               2011             634
               2012             734
               2013             671
               2014             730
999.0 - BLD 04 2010             243
               2011             302
               2012             273
               2013             365
               2014             365
999.0 - BLD 05 2010             275
               2011              59
               2012             123
               2013             365
               2014             333
999.0 - BLD 06 2010             334
               2011             333
               2012             367
               2013             271
               2014             273
999.0 - BLD 07 2010             635
               2011             634
               2012             550
               2013             730
               2014             671
999.0 - BLD 08 2010             364
               2011             365
               2012             302
               2013             333
               2014             365

[16855 rows x 1 columns]

**Many buildings seem to have more than 365 days recorded in a calendar year, which is due to multiple meter readings for the same time period. Have to investigate how to deal with these data anomalies **